### PR TITLE
[Inductor][CPU] Fuse SmoothQuant int8 linear pattern

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #142015
* __->__ #139595

**About the PR**
In the implementation of SmoothQuant in Torchao, quantized linear is computed by `_int_mm(a, b)` + `mul(b_scale)` + `mul(a_scale)` (+ optional `add` for bias) with `reshape` and `convert_dtype` in between.
This PR adds a pass to fuse the corresponding patterns:
- (no bias) `reshape -> _int_mm -> convert_element_type -> (expand -> mul) -> mul -> reshape`
- (with bias) `pattern_no_bias -> add -> reshape -> reshape`

The patterns are replaced by `onednn.qlinear_pointwise` and `onednn.qlinear_prepack`, the latter of which is evaluated and frozen during the freezing process of Inductor. The final graph contains `onednn.qlinear_pointwise` only with packed weight constants.

Note that `onednn.qlinear_pointwise` does not support per-channel quantization of activation, which is a limitation of oneDNN library, so in that case we set activation scale to 1 and bias to none and apply scales and add bias after `onednn.qlinear_pointwise`.

**Validation results**
Accuracy/perplexity is not changed with or without this fusion pass.
Latency is improved by >10% with the fusion pass.
Test method:
- Model: EleutherAI/gpt-j-6b
- Hardware: Intel(R) Xeon(R) Platinum 8490H, running on 1 socket, 60 cores
- Using Intel OMP and Tcmalloc
- Running [the example script of SmoothQuant in Torchao](https://github.com/pytorch/ao/blob/main/torchao/prototype/smoothquant/example.py) with `TORCHINDUCTOR_FREEZING=1 numactl -N1 python example.py -m EleutherAI/gpt-j-6b --device=cpu --quant-mode=dynamic --compile`

**Test plan**
```
python test/inductor/test_mkldnn_pattern_matcher.py -k test_smooth_quant_with_int_mm
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D65702807](https://our.internmc.facebook.com/intern/diff/D65702807)

### CODE REVIEW

**Comment by vadimkantorov** [comment]
A bit related on adding a core frontend / ux function for such scaled float<>int conversions (adding a frontend function is useful to promote safe, correct, fast/optimizable idioms):
- https://github.com/pytorch/pytorch/issues/106624

**Comment by Xia-Weiwen** [comment]
> A bit related (without mm) on adding a core frontend / ux function for such scaled, saturated float<>int casts (adding a frontend function is useful to promote safe, correct, fast/optimizable idioms and maybe also minimize the needed number of manual fusable patterns with mms):
> 
> * [[feature request] [ux] Frontend methods for fused  elementwise affine transform: mul+add+dtype convert + support  `integer_tensor.mul_(float_constant)` and `float_tensor.mul(some_constant, out = integer_tensor)` maybe via new args `rounding_mode=...` and `dtype=...` + maybe support OpenCV-style saturated dtype conversions (e.g. `clamp_` before conversion) #106624](https://github.com/pytorch/pytorch/issues/106624)

Hi. Looks like you are looking for a frontend API for `mul - add - clamp - convert`? However, this PR is not intended for end users. And it's specific to the graph pattern of SmoothQuant from Torchao.

File: torch/_inductor/fx_passes/quantization.py [hunk 1]
```diff
@@ -2529,6 +2529,190 @@ def _register_qlinear_weight_prepack():
...
 
 
+def _register_smooth_quant_int_mm_pattern():
+    """
+    The pattern is:
+      (no bias) reshape -> _int_mm -> convert_element_type -> (expand -> mul) -> mul -> reshape
+    or
+      (with bias) pattern_no_bias -> add -> reshape -> reshape
+    """
+    pattern_no_bias = CallFunction(
```
Comment (jerryzh168) [inline torch/_inductor/fx_passes/quantization.py:2539]: wondering if https://github.com/pytorch/pytorch/pull/139102 will be able to help simplify the pattern as well

File: torch/_inductor/fx_passes/quantization.py [hunk 2]
```diff
@@ -2529,6 +2529,190 @@ def _register_qlinear_weight_prepack():
...
+                KeywordArg("bias"),
+            ),
+            Arg(),
+        ),
+        KeywordArg("out_shape_with_bias"),
+    )
+
+    def _validate_pattern(match: Match):
+        for n in match.nodes:
+            if len(n.users) != 1 and n != match.output_node():
```
Comment (leslie-fang-intel) [inline torch/_inductor/fx_passes/quantization.py:2584]: I think it's not needed of `len(n.users) != 1`. The pattern matcher should cover this check.

File: torch/_inductor/fx_passes/quantization.py (hunk 1 above)
Comment (Xia-Weiwen) [reply torch/_inductor/fx_passes/quantization.py:2539]: Thanks for the info. Can `InvokeQuant` represent the pattern of quantization or any pattern?

**Review by leslie-fang-intel — APPROVED** [review]
Overall LGTM

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 3]
```diff
@@ -1155,10 +1157,11 @@ namespace at::native {
       torch::List<std::optional<at::Scalar>> unary_post_op_args,
       c10::string_view unary_post_op_algorithm) {
 #if AT_MKLDNN_ENABLED()
-    TORCH_CHECK(act_scale.numel() == 1 && act_zero_point.numel() == 1,
-        "onednn int8 linear: act scale/zp size should be 1");
+    TORCH_CHECK(act_scale.numel() == 1 && act_zero_point.numel() <= 1,
+        "onednn int8 linear: act scale/zp size should be 1/<=1");
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp:1163]: Thanks. Added.

File: torch/_inductor/fx_passes/quantization.py (hunk 2 above)
Comment (Xia-Weiwen) [reply torch/_inductor/fx_passes/quantization.py]: Thanks. Removed.

**Review by jgong5 — APPROVED** [review]

File: torch/_inductor/fx_passes/quantization.py (hunk 1 above)
Comment (jerryzh168) [reply torch/_inductor/fx_passes/quantization.py:2539]: it's mostly targeting dequant I think, see: https://github.com/pytorch/pytorch/pull/139102#discussion_r1821576857

Comment (Xia-Weiwen) [reply torch/_inductor/fx_passes/quantization.py:2539]: I see. Do you suggest waiting for that PR landing first then using `InvokeQuant` in this PR?

**Review by jerryzh168 — APPROVED** [review]

File: torch/_inductor/fx_passes/quantization.py (hunk 1 above)
Comment (Xia-Weiwen) [reply torch/_inductor/fx_passes/quantization.py:2539]: Got it. Thanks.

**Comment by ZainRizvi** [comment]
@pytorchbot revert -c ghfirst -m "Sorry but the newly added tests in test_mkldnn_pattern_matcher.py fail internally. See D65661038 for more details"

@jerryzh168 can you please help this PR get landed internally?

Error looks like:

```
AssertionError: Scalars are not equal!

Expected 10 but got 7.
Absolute difference: 3
Relative difference: 0.3
```

**Comment by Xia-Weiwen** [comment]
Hi @ZainRizvi @jerryzh168 May I know how to repro the failure? It looks good in public CI. Are there any changes in Inductor internally that change the graph after aot_autograd? Thanks.

**Comment by jerryzh168** [comment]
seems like a numerics error:

```

======================================================================
FAIL: test_smooth_quant_with_int_mm (caffe2.test.inductor.test_mkldnn_pattern_matcher.TestPatternMatcher)
This testcase check if we can match the SmoothQuant int8 linear pattern from Torchao.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/xarfuse/uid-30041/00f1c470-seed-nspid4026535118_cgpid7058806-ns-4026535106/torch/testing/_internal/common_quantization.py", line 413, in wrapper
    fn(*args, **kwargs)
  File "/mnt/xarfuse/uid-30041/00f1c470-seed-nspid4026535118_cgpid7058806-ns-4026535106/torch/testing/_internal/common_quantization.py", line 365, in wrapper
    fn(*args, **kwargs)
  File "/mnt/xarfuse/uid-30041/00f1c470-seed-nspid4026535118_cgpid7058806-ns-4026535106/caffe2/test/inductor/test_mkldnn_pattern_matcher.py", line 2908, in test_smooth_quant_with_int_mm
    self._test_common(
  File "/mnt/xarfuse/uid-30041/00f1c470-seed-nspid4026535118_cgpid7058806-ns-4026535106/caffe2/test/inductor/test_mkldnn_pattern_matcher.py", line 204, in _test_common
    matcher_check_fn()
  File "/mnt/xarfuse/uid-30041/00f1c470-seed-nspid4026535118_cgpid7058806-ns-4026535106/caffe2/test/inductor/test_mkldnn_pattern_matcher.py", line 2903, in matcher_check_fn
    self.assertEqual(
  File "/mnt/xarfuse/uid-30041/00f1c470-seed-nspid4026535118_cgpid7058806-ns-4026535106/torch/testing/_internal/common_utils.py", line 3965, in assertEqual
    raise error_metas.pop()[0].to_error(
AssertionError: Scalars are not equal!
```

**Comment by Xia-Weiwen** [comment]
> seems like a numerics error:

According to @ZainRizvi , it's not numeric issue of computation. In the pattern match pass, we expect 10 nodes are matched but got 7. I suspect there are changes in Inductor so we got different patterns in public CI and internal checks. Are you aware of such changes?
I have just updated this PR and added a print to print graphs in the UT. Could you please import the patch and show me the logs? Thanks.

**Comment by jerryzh168** [comment]
> > seems like a numerics error:
> 
> According to @ZainRizvi , it's not numeric issue of computation. In the pattern match pass, we expect 10 nodes are matched but got 7. I suspect there are changes in Inductor so we got different patterns in public CI and internal checks. Are you aware of such changes? I have just updated this PR and added a print to print graphs in the UT. Could you please import the patch and show me the logs? Thanks.

oh OK, yeah it's a match issue

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65702807).

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 Please let me know if you get the logs. Thanks!

**Comment by jerryzh168** [comment]
> Hi @jerryzh168 Please let me know if you get the logs. Thanks!

I didn't see the match issue anymore actually

**Comment by Xia-Weiwen** [comment]
> > Hi @jerryzh168 Please let me know if you get the logs. Thanks!
> 
> I didn't see the match issue anymore actually

Do you mean the issue is gone? Sounds great! I will remove the print and probably I can merge it again. Thanks.

**Comment by jerryzh168** [comment]
> > > Hi @jerryzh168 Please let me know if you get the logs. Thanks!
> > 
> > 
> > I didn't see the match issue anymore actually
> 
> Do you mean the issue is gone? Sounds great! I will remove the print and probably I can merge it again. Thanks.

yeah looks like it's gone

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 I cannot merge this PR since you have imported it. Could you please help merge it ? Thanks.

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 I cannot merge this PR since you have imported it. Could you please help merge it ? Thanks.

**Comment by jerryzh168** [comment]
> Hi @jerryzh168 I cannot merge this PR since you have imported it. Could you please help merge it ? Thanks.

there is some problem when I'm trying to import the PR, let me try to rebase first

**Comment by Xia-Weiwen** [comment]
@jerryzh168 Thanks.

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65702807).

**Comment by jerryzh168** [comment]
saw this error internally:

```
======================================================================
FAIL: test_smooth_quant_with_int_mm (caffe2.test.inductor.test_mkldnn_pattern_matcher.TestPatternMatcher)
This testcase check if we can match the SmoothQuant int8 linear pattern from Torchao.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/xarfuse/uid-30041/5876e8e8-seed-nspid4026534796_cgpid5308773-ns-4026534791/torch/testing/_internal/common_quantization.py", line 413, in wrapper
    fn(*args, **kwargs)
  File "/mnt/xarfuse/uid-30041/5876e8e8-seed-nspid4026534796_cgpid5308773-ns-4026534791/torch/testing/_internal/common_quantization.py", line 365, in wrapper
    fn(*args, **kwargs)
  File "/mnt/xarfuse/uid-30041/5876e8e8-seed-nspid4026534796_cgpid5308773-ns-4026534791/caffe2/test/inductor/test_mkldnn_pattern_matcher.py", line 2908, in test_smooth_quant_with_int_mm
    self._test_common(
  File "/mnt/xarfuse/uid-30041/5876e8e8-seed-nspid4026534796_cgpid5308773-ns-4026534791/caffe2/test/inductor/test_mkldnn_pattern_matcher.py", line 204, in _test_common
    matcher_check_fn()
  File "/mnt/xarfuse/uid-30041/5876e8e8-seed-nspid4026534796_cgpid5308773-ns-4026534791/caffe2/test/inductor/test_mkldnn_pattern_matcher.py", line 2903, in matcher_check_fn
    self.assertEqual(
  File "/mnt/xarfuse/uid-30041/5876e8e8-seed-nspid4026534796_cgpid5308773-ns-4026534791/torch/testing/_internal/common_utils.py", line 3977, in assertEqual
    raise error_metas.pop()[0].to_error(
AssertionError: Scalars are not equal!

Expected 10 but got 7.
Absolute difference: 3
Relative difference: 0.3
```

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168  Thanks for the results. It's still about pattern match. I have added some prints in the code. Could you please import it again and show me the log before the error? Thank!

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65702807).

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 4]
```diff
@@ -2959,6 +2960,95 @@ def matcher_check_fn():
...
+            [torch.float, torch.bfloat16]
+            if torch.ops.mkldnn._is_mkldnn_bf16_supported()
+            else [torch.float]
+        )
+        per_channel_list = [True, False]
+        dynamic_list = [True, False]
+        for has_bias, dtype, per_channel_quant, dynamic in itertools.product(
+            has_bias_list, dype_list, per_channel_list, dynamic_list
+        ):
+            print("[info] ===== case:", has_bias, dtype, per_channel_quant, dynamic, "=====")
```
Comment (jerryzh168) [inline test/inductor/test_mkldnn_pattern_matcher.py:3027]: it doesn't error out anymore after you add this print somehow, maybe you can rewrite this test with parametrize like https://github.com/pytorch/pytorch/blob/bae951030752f2d7f06c15af0c59e54026d851e9/test/quantization/pt2e/test_quantize_pt2e.py#L1178-L1179

Comment (Xia-Weiwen) [reply test/inductor/test_mkldnn_pattern_matcher.py]: Thanks. I have updated the PR. Please help import again. Thanks.

**Comment by sanchitintel** [comment]
Hi @Xia-Weiwen @leslie-fang-intel, this pattern doesn't use GEMM auto-tuning if GEMM autotuning configs are enabled. I'll add its support in another PR. Thanks!

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65702807).

**Comment by jerryzh168** [comment]
multiple things failing: 
```
caffe2/test/inductor:mkldnn_pattern_matcher_cpu - test_smooth_quant_with_int_mm_has_bias_False_bfloat16_per_channel_quant_True_dynamic_True (caffe2.test.inductor.test_mkldnn_pattern_matcher.TestPatternMatcher)

caffe2/test/inductor:mkldnn_pattern_matcher_cpu - test_smooth_quant_with_int_mm_has_bias_False_float32_per_channel_quant_True_dynamic_True (caffe2.test.inductor.test_mkldnn_pattern_matcher.TestPatternMatcher)

caffe2/test/inductor:mkldnn_pattern_matcher_cpu - test_smooth_quant_with_int_mm_has_bias_True_bfloat16_per_channel_quant_True_dynamic_True (caffe2.test.inductor.test_mkldnn_pattern_matcher.TestPatternMatcher)

caffe2/test/inductor:mkldnn_pattern_matcher_cpu - test_smooth_quant_with_int_mm_has_bias_True_float32_per_channel_quant_True_dynamic_True (caffe2.test.inductor.test_mkldnn_pattern_matcher.TestPatternMatcher)
```

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 Thanks for your feedback. So the issue is still there. Adding prints was not helpful. Do you have any suggestions? Thanks.

**Comment by jerryzh168** [comment]
@Xia-Weiwen @sanchitintel there are some fbcode only options enabled in torch.compile I think, will need to take a closer look to see what is happening

File: torch/_inductor/fx_passes/quantization.py [hunk 5]
```diff
@@ -2529,6 +2529,226 @@ def _register_qlinear_weight_prepack():
...
+                        [],  # post op args
+                        "",  # post op algorithm
+                    )
+                    new_linear_node = match.graph.call_function(
+                        torch.ops.onednn.qlinear_pointwise, args=new_args
+                    )
+                    # apply x scale
+                    new_out_node = match.graph.call_function(
+                        aten.mul.Tensor, args=(new_linear_node, x_scale)
+                    )
```
Comment (sanchitintel) [inline torch/_inductor/fx_passes/quantization.py:2724]: Is there any way to add application of activation's scale as a `mul` post-op for oneDNN qlinear instead? Inductor can't fuse it as an epilogue. Or was this approach chosen because using a `mul` post-op is slower? Thanks!

Comment (Xia-Weiwen) [reply torch/_inductor/fx_passes/quantization.py:2724]: Mul as a post op may be an option. We didn't do that because we haven't implemented qlinear-mul yet. https://github.com/Xia-Weiwen/pytorch/blob/2398e758d2493c8142d4b7f9c72207f05991e9e9/aten/src/ATen/native/quantized/cpu/qlinear.cpp#L914

Comment (sanchitintel) [reply torch/_inductor/fx_passes/quantization.py:2724]: Thanks for the info!

**Comment by jerryzh168** [comment]
actually @Xia-Weiwen @sanchitintel is this needed for internal use cases? if not we can just skip in fbcode for now

**Comment by Xia-Weiwen** [comment]
> actually @Xia-Weiwen @sanchitintel is this needed for internal use cases? if not we can just skip in fbcode for now

I don't know if it's internal or not 😂 It would help anyone who wants to run SmoothQuant with Torchao and Inductor on CPU. Anyway, skipping it fbcode sounds a good idea. I will skip it if it's OK to you. Thanks.

**Comment by jerryzh168** [comment]
> > actually @Xia-Weiwen @sanchitintel is this needed for internal use cases? if not we can just skip in fbcode for now
> 
> I don't know if it's internal or not 😂 It would help anyone who wants to run SmoothQuant with Torchao and Inductor on CPU. Anyway, skipping it fbcode sounds a good idea. I will skip it if it's OK to you. Thanks.

yeah I think it's fine for now, I haven't heard a meta internal use case for it right now. also I don't know when I'll have time to look into the fbcode flag issue.

**Comment by jerryzh168** [comment]
please skip and add a comment for this

**Comment by Xia-Weiwen** [comment]
> please skip and add a comment for this

Thanks. It's added. Please take a look.

**Comment by sanchitintel** [comment]
Hi @Xia-Weiwen, when I run the TorchAO UT (`python test/integration/test_integration.py -v -k test_non_dynamically_quantizable_linear`) for smoothquant by modifying it to run for CPU instead of CUDA, I don't see the pattern getting matched.

**Comment by sanchitintel** [comment]
TorchAO UT `test/integration/test_integration.py -v -k SmoothquantIntegrationTest.test_non_dynamically_quantizable_linear` has a slightly different pattern. I opened #141851 for it by using ghstack to include this PR's changes.

Thanks!

File: torch/_inductor/fx_passes/quantization.py [hunk 6]
```diff
@@ -2539,6 +2539,226 @@ def _register_qlinear_weight_prepack():
...
+                pattern_no_bias_2,
+                KeywordArg("bias"),
+            ),
+            Arg(),
+        ),
+        KeywordArg("out_shape_with_bias"),
+    )
+
+    def _validate_pattern(match: Match):
+        if len(match.nodes) not in [6, 7, 10]:
```
Comment (leslie-fang-intel) [inline torch/_inductor/fx_passes/quantization.py:2611]: Do we need to check `kwargs["b"]` is a constant on the graph?

Comment (sanchitintel) [reply torch/_inductor/fx_passes/quantization.py:2611]: Probably not, since the pattern that's registered is a freezing pattern.

Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/quantization.py:2611]: > registered is a freezing pattern

Per my understanding, it only means these patterns work when user turns on freezing. But it doesn't mean the second input node to `torch.int_mm` is a constant to do weight prepack.
It may not cause accuracy issue, but I am afraid insert weight prepack node when second input of `torch.int_mm` is not constant seems redundant.

Comment (sanchitintel) [reply torch/_inductor/fx_passes/quantization.py:2611]: Freezing wouldn't be used with use-cases such as QAT, in which the second input of `torch._int_mm` may not be a constant (the API is not user-facing and is primarily used to replace `linear`s while quantizing a model), so in practice, that scenario may not arise.

But then, in the future, users may use `torch._int_mm` API with some new use-case (may replace a matmul that wasn't a linear) & may encounter the scenario you mentioned with freezing enabled, so perhaps it's better to add a constant-check.

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 7]
```diff
@@ -3023,6 +3025,93 @@ def matcher_check_fn():
...
+                c = c * self.b_scale
+                if self.has_bias:
+                    c = c.reshape([1, *list(c_shape)])
+                    c = c + self.bias
+                    c = c.reshape(c_shape)
+                c = c.reshape(out_shape)
+                return c
+
+        mod = Mod(dtype, has_bias, per_channel_quant).eval()
+        a = torch.randint(q_min, q_max, [1, M, in_feature], dtype=torch.int8)
```
Comment (sanchitintel) [inline test/inductor/test_mkldnn_pattern_matcher.py:3092]: @leslie-fang-intel, here, 1 is necessary for the UTs where `dynamic` is `False`.
@Xia-Weiwen, can you please elaborate on why? Thanks

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 8]
```diff
@@ -3023,6 +3025,93 @@ def matcher_check_fn():
...
+    @parametrize("dtype", [torch.float, torch.bfloat16])
+    @parametrize("per_channel_quant", [True, False])
+    @parametrize("dynamic", [True, False])
+    def test_smooth_quant_with_int_mm(
+        self, has_bias, dtype, per_channel_quant, dynamic
+    ):
+        r"""
+        This testcase check if we can match the SmoothQuant int8 linear pattern from Torchao.
+        The pattern is:
+            (no bias) reshape -> _int_mm -> convert_element_type -> (expand -> mul) -> mul -> reshape
```
Comment (sanchitintel) [inline test/inductor/test_mkldnn_pattern_matcher.py:3042]: @leslie-fang-intel, same comment here as https://github.com/pytorch/pytorch/pull/141851#discussion_r1866946236?

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 7 above)
Comment (Xia-Weiwen) [reply test/inductor/test_mkldnn_pattern_matcher.py:3092]: @sanchitintel Sorry I didn't get your point. Could you rephrase your question?

File: torch/_inductor/fx_passes/quantization.py (hunk 6 above)
Comment (Xia-Weiwen) [reply torch/_inductor/fx_passes/quantization.py:2611]: So, do we need to check b is constant? Thanks.

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 8 above)
Comment (Xia-Weiwen) [reply test/inductor/test_mkldnn_pattern_matcher.py:3042]: @sanchitintel What pattern did you see?

Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:3042]: @leslie-fang-intel's concern seems to be that this visual depiction indicates that the output of `torch._int_mm` is fed into `expand`

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 7 above)
Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:3092]: @Xia-Weiwen, if I use a 2D shape `[M, in_feature]` for `a`, then all UTs pass if `dynamic_shape` is `False`.
But the UTs corresponding to `dynamic_shape` being `True` fail in that case.

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 8 above)
Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:3042]: @Xia-Weiwen, I now think that @leslie-fang-intel was asking me to include `expand -> mul` within brackets, just like you have.

@leslie-fang-intel, also, why not use indicate here that `expand` is optional?

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 7 above)
Comment (Xia-Weiwen) [reply test/inductor/test_mkldnn_pattern_matcher.py:3092]: Yes. I think in real cases `a` is 3D, like `[batch, seq_len, hidden]`. If `a` is 2D, the pattern would be different I think.

Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:3092]: Thanks for the info, @Xia-Weiwen! :)

File: torch/_inductor/fx_passes/quantization.py (hunk 6 above)
Comment (Xia-Weiwen) [reply torch/_inductor/fx_passes/quantization.py:2611]: I have added a check. Thanks.

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 I have updated this PR according to latest comments, including skipping the UT in fbcode. If everything looks good to you, could you please import it again? Thanks.

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 8 above)
Comment (Xia-Weiwen) [reply test/inductor/test_mkldnn_pattern_matcher.py:3131]: Ok

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 7 above)
Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:3181]: @Xia-Weiwen, thanks for the info! #141851 will handle that case.

**Comment by jerryzh168** [comment]
@Xia-Weiwen I can't import the PR, even after rebase, it is saying: `error: patch failed: fbcode/caffe2/test/inductor/test_mkldnn_pattern_matcher.py:174
error: fbcode/caffe2/test/inductor/test_mkldnn_pattern_matcher.py: patch does not apply`

**Comment by Xia-Weiwen** [comment]
@jerryzh168 Thanks for the info. I have rebased this PR before I added this comment https://github.com/pytorch/pytorch/pull/139595#issuecomment-2514071824 actually. Let me try again.

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 I have rebased manually. Please try again. Thanks.

**Comment by jerryzh168** [comment]
@Xia-Weiwen thanks, still can't do it, maybe you can create a new PR and I can try again

**Comment by Xia-Weiwen** [comment]
> @Xia-Weiwen thanks, still can't do it, maybe you can create a new PR and I can try again

Hi @jerryzh168 I have created a new PR: https://github.com/pytorch/pytorch/pull/142036 Please have a try, thanks.
